core-api/src/main/java/com/optimizely/ab/internal/ConditionUtils.java

```diff
@@ -0,0 +1,207 @@
+                conditions.add(new UserAttribute((String)conditionMap.get("name"), (String)conditionMap.get("type"),
+                        (String)conditionMap.get("match"), conditionMap.get("value")));
+            }
+        }
+
+        Condition condition;
+        switch (operand) {
+            case "and":
+                condition = new AndCondition(conditions);
+                break;
+            case "or":
+                condition = new OrCondition(conditions);
+                break;
+            case "not":
+                condition = new NotCondition(conditions.get(0));
+                break;
+            default:
+                condition = new OrCondition(conditions);
+                break;
+        }
+
+        return condition;
+    }
+
+    /**
+     * Parse conditions from org.json.JsonArray
+     * @param conditionJson jsonArray to parse
+     * @return condition parsed from conditionJson.
+     */
+    static public Condition parseConditions(org.json.JSONArray conditionJson) {
+        List<Condition> conditions = new ArrayList<Condition>();
+        String operand = (String)conditionJson.get(0);
+
+        for (int i = 1; i < conditionJson.length(); i++) {
+            Object obj = conditionJson.get(i);
+            if (obj instanceof org.json.JSONArray) {
+                conditions.add(parseConditions((org.json.JSONArray) conditionJson.get(i)));
+            } else if (obj instanceof String) {
+                conditions.add(new AudienceIdCondition((String)obj));
+            } else {
+                org.json.JSONObject conditionMap = (org.json.JSONObject)obj;
+                String match = null;
+                Object value = null;
+                if (conditionMap.has("match")) {
+                    match = (String) conditionMap.get("match");
+                }
+                if (conditionMap.has("value")) {
+                    value = conditionMap.get("value");
+                }
+                conditions.add(new UserAttribute((String)conditionMap.get("name"), (String)conditionMap.get("type"),
+                        match, value));
+            }
+        }
+
+        Condition condition;
+        switch (operand) {
+            case "and":
+                condition = new AndCondition(conditions);
+                break;
+            case "or":
+                condition = new OrCondition(conditions);
+                break;
+            case "not":
+                condition = new NotCondition(conditions.get(0));
+                break;
+            default:
+                condition = new OrCondition(conditions);
+                break;
+        }
+
+        return condition;
+    }
+
+    /**
+     * resolveAudienceIdConditions walks the conditions list and fills in the audience in any AudienceIdConditions by
+     * looking up audience by id and setting the audience in the AudienceIdCondition.  It uses recursion to walk the list and find
+     * any AudienceIdCondition not set.
```

Comment:
Got it, thanks!

I do think it might be simpler to have `AudienceIdCondition.evaluate` look up its audience by ID, rather than requiring the mapping to have been performed (with its own tree traversal logic) in a priori step. I think the temporal dependency here is kinda spaghetti - it's another way things go wrong and I'm not seeing the benefit (though I might just be missing it)

Reply:
I agree that there might be a better way to do this.  We could attempt to resolve during parsing.  This would require that either audiences are parsed first or the resolution happens at project config parse time.  The point is that there is no easy way to get the project config to resolve the audience id at the time of parsing (you are actually in the middle of parsing the project config at this time)